WebDiscover: Partially implement auto deploy database server view

kimlisa opened the pull request:
part of https://github.com/gravitational/teleport/issues/22132

Note: this PR will not enable the auto deploy yet, a second PR will be opened to finalize this view for the `configure command script`

- Adds a new screen that allows user to `auto deploy` db agent. This will be the default screen after user is finished selecting an AWS integration name
- User can switch back to the `manual deploy` method if they want


<details>
<summary>
Screenshots
</summary>

initial state

<img width="1672" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/89932bab-0508-45b3-a4d0-d1eb4d30b3fc">

error requesting to deploy

<img width="1021" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/85640b66-4ff7-44bc-8e64-5a7462440495">

waiting to deploy

<img width="1043" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/d5127e92-8c25-4aa8-ba58-bf87f5623d98">

waiting too long to deploy

<img width="940" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/ebcb806f-17e1-4231-86f2-7984609fd516">



</details>

kimlisa committed:
Add new endpoint, types, update types

kimlisa committed:
Preserve user selected aws rds db

kimlisa committed:
Rename DownloadScript to ManualDeploy

- Remove duplicate test
- Use discover context instead of passed in props
- Rename util to common
- Move hasMatchingLabels to common.tsx for re-use

kimlisa committed:
Add another action button and pass autoFocus field

kimlisa committed:
Implement auto deploy screen

kimlisa committed:
Implement auto manual view toggler

kimlisa committed:
Update db view config

kimlisa committed:
Only allow back on user trait if user manually deployed agent

kimlisa committed:
Add TODO comment

kimlisa commented:
@ryanclark @ibeckermayer this is ready for review. 

the only thing that's not set in stone are copies and what the command script will look like and what user input the script requires (so far just the role task name input for sure)

the only way to test this flow atm is by manually setting up the roles and policies required for the `deploy` button to work, let me know if interested

ibeckermayer commented on web/packages/teleport/src/Discover/Database/DeployService/DeployService.tsx:
```diff
@@ -0,0 +1,38 @@
+  return <ManualDeploy toggleDeployMethod={toggleDeployMethod} />;
+}
+
+export type DeployServiceProp = {
+  toggleDeployMethod(): void;
+};
```
Everywhere this is used it's used like `{ toggleDeployMethod }: Partial<DeployServiceProp>`, you can remove the `Partial` and change it to
```suggestion
export type DeployServiceProp = {
  toggleDeployMethod?(): void;
};
```

ibeckermayer commented on web/packages/teleport/src/Discover/Database/DeployService/AutoDeploy/AutoDeploy.tsx:
```diff
@@ -0,0 +1,364 @@
+            />
+
+            {/* step two */}
+            <StyledBox mb={5}>
+              <Text bold>Step 2</Text>
+              <Box mb={4}>
+                <Labels
+                  labels={labels}
+                  setLabels={setLabels}
+                  disableBtns={attempt.status === 'processing'}
+                  showLabelMatchErr={showLabelMatchErr}
+                  dbLabels={dbLabels}
+                  autoFocus={false}
+                />
+              </Box>
+              <ButtonSecondary
+                width="215px"
+                type="submit"
+                onClick={() => handleDeploy(validator)}
+                disabled={attempt.status === 'processing'}
+                mb={2}
+              >
+                Deploy Teleport Service
+              </ButtonSecondary>
+              {hasError && (
+                <Box>
+                  <TextIcon mt={3}>
+                    <Icons.Warning ml={1} color="error.main" />
+                    Encountered Error: {attempt.statusText}
+                  </TextIcon>
+                </Box>
+              )}
+            </StyledBox>
```
Could be wrapped into it's own `DefineMatcherLabels` component

kimlisa replied:
i could, but didn't think it was so long that it was unreadable so left it as is (there were so many props to pass). if you feel strongly about it i'll pull it out

kimlisa commented on web/packages/teleport/src/Discover/Database/DeployService/AutoDeploy/AutoDeploy.tsx:
```diff
@@ -0,0 +1,364 @@
+    <>
+      <Header>Automatically Deploy a Database Service</Header>
+      <HeaderSubtitle>
+        Teleport needs an agent to be able to connect to your database. With a
+        few permission configurations, Teleport can spin up an ECS Fargate
```
replaced agent with `database service` i think agent is an old term (but we internally still use it)

kimlisa commented on web/packages/teleport/src/Discover/Database/common.tsx:
```diff
@@ -0,0 +1,190 @@
+        {!hasDbLabels && (
+          <>
+            Since no labels were defined for the registered database from the
+            previous step, the matcher labels are defaulted to wildcards which
```
i didn't think this was necessary since we show the user (it'll just be readonly so they can't change it for this case):

<img width="613" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/e014a66d-cf3d-4aa9-92d4-7e370eac0a57">

kimlisa commented on web/packages/teleport/src/Discover/Database/common.tsx:
```diff
@@ -0,0 +1,190 @@
+        )}
+        {hasDbLabels && (
+          <>
+            Default wildcards label allows this database service to match any
+            database.
```
i modified copies so far like this:

<img width="831" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/955d766d-a1fb-4072-abe6-8219a3a8434d">

<img width="1024" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/5e9b70d0-3598-45e9-8516-a42fbd06ccb3">

<img width="927" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/bb461c08-5ada-474e-9999-b97e81650718">

kimlisa commented on web/packages/teleport/src/Discover/Shared/LabelsCreater/LabelsCreater.tsx:
```diff
@@ -29,12 +29,14 @@ export function LabelsCreater({
   disableBtns?: boolean;
   isLabelOptional?: boolean;
   noDuplicateKey?: boolean;
+  autoFocus?: boolean;
```
actually that should be false, thanks

kimlisa committed:
Address CRs

kimlisa commented on web/packages/teleport/src/services/databases/makeDatabase.ts:
```diff
@@ -16,19 +16,21 @@ limitations under the License.
 
   // Only setting RDS fields for now.
-  let rds;
+  let awsRds: Aws;
```
i just mimicked the type that was returned to me,  we get this [AWS](https://github.com/gravitational/teleport/blob/5ba028bc5843db6506f4d41919221286ce60bb40/api/proto/teleport/legacy/types/types.proto#L388) type from the back (which has a rds field in it)

i think it makes sense to leave it as is for now, a non-empty `aws` field lets us know that this db is hosted by aws, and rds field describes what type of rds.

ibeckermayer replied:
Got it. I'm not actually sure about this:

```proto
  // RDS contains RDS specific metadata.
  RDS RDS = 3 [
    (gogoproto.nullable) = false,
    (gogoproto.jsontag) = "rds,omitempty"
  ];
```

If we have `nullable = false` but also `omitempty`, will it ever come back empty?

kimlisa replied:
yes, for the web api, it will return the field as `undefined` b/c we check if it's aws hosted before setting the field:
https://github.com/gravitational/teleport/blob/b3e83faead6527fb25c4d5b7f04da711757c6606/lib/web/ui/server.go#L281

ibeckermayer replied:
I see that that answers that `aws` might be `undefined` when being sent back via the API, my question is whether `aws.rds` itself can ever be empty. It seems like it can't (even though we have `omitempty`), I think it will always be part of the `aws` object if that's returned, it will just have go's default values.

Ultimately what I'm asking is whether we should remove the optionality of `rds` from the `Aws` object, making it:

```typescript
export type Aws = {
  rds: Pick<AwsRdsDatabase, 'resourceId' | 'region' | 'subnets'>;
};
```

I don't see anywhere that it's being treated as `undefined`.

kimlisa replied:
oh i get what you're saying now, yes you're completely right. edited: https://github.com/gravitational/teleport/pull/27924/commits/874231b0cb9086c76739dced5dfdcd88a04e02ec

ryanclark commented on web/packages/teleport/src/Discover/Database/DeployService/DeployService.tsx:
```diff
@@ -0,0 +1,38 @@
+import { AutoDeploy } from './AutoDeploy';
+
+export function DeployService() {
+  const [deployMethod, setDeployMethod] = useState<'manual' | 'auto'>('auto');
```
`'manual' | 'auto'` is used a couple of times, should it be moved to its own type?

kimlisa replied:
sorry i wasn't sure where the other one was 😅

ryanclark commented on web/packages/teleport/src/Discover/Database/DeployService/ManualDeploy/ManualDeploy.story.tsx:
```diff
@@ -0,0 +1,123 @@
+        <FeaturesContextProvider value={[]}>
+          <DiscoverProvider mockCtx={discoverCtx}>
+            <PingTeleportProvider
+              interval={props.interval || 100000}
```
Thoughts on moving `100000` to a constant? `const DEFAULT_PING_INTERVAL = 1000 * 100; // 100 seconds`

kimlisa committed:
Fix lint/test

kimlisa committed:
Address CRs

kimlisa committed:
Revert back to original flow

kimlisa commented:
@ibeckermayer @ryanclark i decided to make a second PR to finish the `command script` work, and merge this PR as is. I reverted back to the original flow so the new `auto deploy` feature will not be enabled.

kimlisa committed:
Define type for service deploy method

ryanclark commented on web/packages/teleport/src/Discover/Database/DeployService/AutoDeploy/AutoDeploy.tsx:
```diff
@@ -0,0 +1,370 @@
+        taskRoleArn,
+        databaseAgentMatcherLabels: labels,
+      })
+      .then(setDeploySvcResp)
```
Should this set the attempt to success?

kimlisa replied:
no it wasn't necessary (save a re-render?) because right after this call succeeds, we move to `pinging` for the db service to pick up the database (which is still a processing state)

i left a clarifying comment

ibeckermayer commented on web/packages/teleport/src/Discover/Database/common.tsx:
```diff
@@ -0,0 +1,207 @@
+            Teleport, leave this for now.
+          </Text>
+          <Text typography="subtitle1" mb={2}>
+            Alternatively, you can define narrower labels that this database
+            service will use to identify the databases you register
```
```suggestion
            Alternatively, you can define narrower labels for
            identifying the databases you register
```

kimlisa committed:
Address CR and revert a change

- aws field for database has to be empty if the
  result comes back empty. this field determines
  if the database is hosted by aws
- add a clarifying comment

kimlisa committed:
Remove optional rds field

kimlisa committed:
Merge branch 'master' into lisa/discover-auto-agent
